packages/evm/contracts/adapters/Router/RouterReporter.sol

```diff
@@ -0,0 +1,111 @@
+}
+
+interface IRouterGasStation {
+    function payFee(string memory destChainId, uint256 destGasLimit) external payable returns (uint256);
+}
+
+contract RouterReporter is Reporter, Ownable {
+    using Strings for uint256;
+
+    string public constant PROVIDER = "router";
+    bytes32 private constant NULL_STRING = keccak256("");
+    IRouterGateway public immutable ROUTER_GATEWAY;
+    IRouterGasStation public immutable ROUTER_GAS_STATION;
+    uint256 public immutable CURRENT_GATEWAY_VERSION;
+    string public feePayer;
+
+    mapping(uint256 => string) public chainIds;
+
+    error ChainIdNotSupported(uint256 chainId);
+    error InsufficientFeePassed();
+
+    event ChainIdSet(uint256 indexed chainId, string indexed chainIdString);
+    event FeePayerSet(string oldFeePayer, string feePayer);
+
+    constructor(
+        address headerStorage,
+        address yaho,
+        address routerGateway,
+        address routerGasStation,
+        string memory routerFeePayer
+    ) Reporter(headerStorage, yaho) {
+        ROUTER_GATEWAY = IRouterGateway(routerGateway);
+        ROUTER_GAS_STATION = IRouterGasStation(routerGasStation);
+
+        feePayer = routerFeePayer;
+
+        CURRENT_GATEWAY_VERSION = ROUTER_GATEWAY.currentVersion();
+        ROUTER_GATEWAY.setDappMetadata(routerFeePayer);
+    }
+
+    function setRouterFeePayer(string memory routerFeePayer) external onlyOwner {
+        string memory oldFeePayer = feePayer;
+        feePayer = routerFeePayer;
+        ROUTER_GATEWAY.setDappMetadata(routerFeePayer);
+
+        emit FeePayerSet(oldFeePayer, routerFeePayer);
+    }
+
+    function setChainIdStringByChainId(uint256 chainId, string calldata chainIdString) external onlyOwner {
+        chainIds[chainId] = chainIdString;
+        emit ChainIdSet(chainId, chainIdString);
+    }
+
+    function getRequestMetadata() internal pure returns (bytes memory) {
+        bytes memory requestMetadata = abi.encodePacked(
+            uint64(200_000),
+            uint64(0),
+            uint64(0),
+            uint64(0),
+            uint128(0),
+            uint8(0),
+            false,
+            string("")
+        );
+        return requestMetadata;
+    }
+
+    function _dispatch(
+        uint256 targetChainId,
+        address adapter,
+        uint256[] memory ids,
+        bytes32[] memory hashes
+    ) internal override returns (bytes32) {
+        string memory targetChainIdStr = chainIds[targetChainId];
+        if (keccak256(abi.encode(targetChainIdStr)) == NULL_STRING) revert ChainIdNotSupported(targetChainId);
+
+        bytes memory payload = abi.encode(ids, hashes);
+        string memory stringAdapter = uint256(uint160(adapter)).toHexString(20);
+        bytes memory requestPacket = abi.encode(stringAdapter, payload);
+        bytes memory requestMetadata = getRequestMetadata();
+
+        uint256 iSendFee = ROUTER_GATEWAY.iSendDefaultFee();
+        if (msg.value < iSendFee) revert InsufficientFeePassed();
```

Comment:
Unfortunately, this would break the integration with `Yaho` because it doesn't handle the distribution of msg.values among adapters. I suggest adding a payable function that accepts ETH so that a user can directly pay the adapter before calling `Yaho.dispatchMessageToAdapters`.

Basically, instead of checking the `msg.value` here, you can check the contract balance:

```sol
if (address(this).balance < iSendFee) revert ....
```